### PR TITLE
Report registration errors from kickstart

### PR DESCRIPTION
**Pull request by M4rtinK**
Show an interactive dialog when a registration attempt triggered
from kickstart fails & give the user option to continue or abort
the installation.

Resolves: rhbz#2000650

### CODE REVIEW

**Comment by M4rtinK** [comment]
rhsm_kickstart_error_v1
Error message improvement suggestions welcome. :-)

rhsm_kickstart_error_quit_v1
Also this is how it looks like when you select Quit - IIRC same as with other early errors.

File: anaconda.py [hunk 1]
```diff
@@ -628,14 +628,24 @@ def _earlyExceptionHandler(ty, value, traceback):
     #   key are available
     from pyanaconda.modules.common.util import is_module_available
     from pyanaconda.modules.common.constants.services import SUBSCRIPTION
+    from pyanaconda.modules.common.errors.subscription import KickstartRegistrationError
+    from pyanaconda.errors import errorHandler, ERROR_RAISE
+
+    def registration_error_handler(message):
+        """Helper function wich raises exception if kickstart triggered registration fails."""
```
Comment (rvykydal) [inline anaconda.py:635]: a typo wich -> which

File: pyanaconda/errors.py [hunk 2]
```diff
@@ -271,6 +271,20 @@ def _insightsErrorHandler(self, exn):
         else:
             return ERROR_RAISE
 
+    def _kickstartRegistrationErrorHandler(self, exn):
+        message = _("An error occurred during registration attempt "
+                    "triggered by the rhsm kickstart command."
+                    "This could have happened due to incorrect rhsm command arguments "
+                    "or subscription infrastructure issues."
```
Comment (rvykydal) [inline pyanaconda/errors.py:278]: Same here.

**Comment by rvykydal** [comment]
> Error message improvement suggestions welcome. :-)

The message looks good to me, I'd just add spaces after periods.

File: anaconda.py [hunk 3]
```diff
@@ -628,14 +628,24 @@ def _earlyExceptionHandler(ty, value, traceback):
     #   key are available
     from pyanaconda.modules.common.util import is_module_available
     from pyanaconda.modules.common.constants.services import SUBSCRIPTION
+    from pyanaconda.modules.common.errors.subscription import KickstartRegistrationError
+    from pyanaconda.errors import errorHandler, ERROR_RAISE
+
+    def registration_error_handler(message):
```
Comment (poncovka) [inline anaconda.py:634]: Please, move this function, for example, to `pyanaconda.ui.lib.subscription`. We are trying to make `anaconda.py` smaller.

File: pyanaconda/modules/common/errors/subscription.py [hunk 4]
```diff
@@ -37,3 +37,9 @@ class UnregistrationError(AnacondaError):
 class SubscriptionError(AnacondaError):
     """Subscription attempt failed."""
     pass
+
+
+@dbus_error("KickstartRegistrationError", namespace=ANACONDA_NAMESPACE)
+class KickstartRegistrationError(AnacondaError):
```
Comment (poncovka) [inline pyanaconda/modules/common/errors/subscription.py:43]: Please, move this error to `pyanaconda.ui.lib.subscription` as well. This exception is not raised by a DBus module, it is just an internal workaround for the UI. Anyway, DBus modules don't raise kickstart-related exceptions. Modules return a kickstart report during parsing a kickstart file and raise more generic errors after that.

**Review by poncovka — CHANGES_REQUESTED** [review]
Otherwise, it looks good to me.

File: anaconda.py (hunk 1 above)
Comment (M4rtinK) [reply anaconda.py:635]: Fixed. :)

File: pyanaconda/errors.py [hunk 5]
```diff
@@ -271,6 +271,20 @@ def _insightsErrorHandler(self, exn):
         else:
             return ERROR_RAISE
 
+    def _kickstartRegistrationErrorHandler(self, exn):
+        message = _("An error occurred during registration attempt "
+                    "triggered by the rhsm kickstart command."
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/errors.py:276]: Oh, right, fixed. :)

File: pyanaconda/errors.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/errors.py:278]: Fixed.

**Comment by M4rtinK** [comment]
> > Error message improvement suggestions welcome. :-)
> 
> The message looks good to me, I'd just add spaces after periods. And maybe a space before '?'.
> 
> And perhaps instead of:
> 
> ```
> Organization unknown_organization does not exist.
> ```
> 
> I'd say:
> 
> ```
> Error detail: Organization unknown_organization does not exist.
> ```
> 
> because just displaying the detail between the question and prompt seems a bit confusing to me ?

Agreed & changed. Thanks for feedback :)

File: anaconda.py (hunk 3 above)
Comment (M4rtinK) [reply anaconda.py]: Sure, moved. :)

File: pyanaconda/modules/common/errors/subscription.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/errors/subscription.py]: Moved as well. BTW, I should keep the `@dbus_error` stuff in place, right ?

Comment (poncovka) [reply pyanaconda/modules/common/errors/subscription.py]: No, you should remove it since it is not a DBus error.

**Comment by M4rtinK** [comment]
PR updated:
rhsm_kickstart_error_quit_v2
IMHO looks much nicer now, thanks for the feedback yet again! :)

File: pyanaconda/errors.py [hunk 6]
```diff
@@ -271,6 +271,20 @@ def _insightsErrorHandler(self, exn):
         else:
             return ERROR_RAISE
 
+    def _kickstartRegistrationErrorHandler(self, exn):
+        message = _("An error occurred during registration attempt "
+                    "triggered by the rhsm kickstart command. "
+                    "This could have happened due to incorrect rhsm command arguments "
+                    "or subscription infrastructure issues. "
+                    "Would you like to ignore this and continue with "
+                    "installation ?")
```
Comment (poncovka) [inline pyanaconda/errors.py:280]: There is an extra space before `?`.

**Review by poncovka — APPROVED** [review]
Thanks!

File: pyanaconda/errors.py (hunk 6 above)
Comment (M4rtinK) [reply pyanaconda/errors.py:280]: An idea from @rvykydal - IMHO it indeed looks better this way.

Comment (poncovka) [reply pyanaconda/errors.py:280]: @VladimirSlavik, what do you think?

Comment (jstodola) [reply pyanaconda/errors.py:280]: jstodola votes for removing the extra space.

Comment (M4rtinK) [reply pyanaconda/errors.py]: Removed. :)

**Review by poncovka — APPROVED** [review]

**Review by rvykydal — APPROVED** [review]

**Pull request merged**